flik/optimize/functions.py

```diff
@@ -0,0 +1,512 @@
+        Finite difference method ("forward" | "central") if str.
+        By default, a central difference Hessian will be used where necessary.
+    tol : float, optional
+        Convergence threshold for gradient norm. Default is 1.0e-9.
+    maxiter : int, optional
+        Maximum number of iterations to perform. Default is 100.
+    method : (Method | str), optional
+        Update method for the (approximated) J(x) or the inverse of J(x).
+        Method name (if str).
+        Otherwise the Method object may be passed directly.
+        Default is Newton method.
+    linesearch : (LineSearch | float | np.ndarray | str), optional
+        Scaling factor for Newton step (if float or np.ndarray).
+        Method name (if str).
+        Otherwise the LineSearch object may be passed directly.
+        Default is no line search.
+    trustregion : (TrustRegion | float | np.ndarray | str), optional
+        Method name (if str).
+        Otherwise the TrustRegion object may be passed directly.
+        Default is no trust region.
+    lsopts : dict, optional
+        Keyword options to line search.
+    tropts : dict, optional
+        Keyword options to trust region.
+    kwargs : dict, optional
+        Keyword options to method.
+
+    Returns
+    -------
+    result : dict
+        A dictionary with the keys:
+        success: Whether or not the optimization succeeded.
+        message: Information about the cause of termination.
+        niter:   Number of actual iterations performed.
+        tol:     Convergence threshold for vector function f norm.
+        x:       Solution vector (root).
+        f:       Objective function evaluated at solution.
+        grad:    Gradient evaluated at solution.
+        H:       Hessian evaluated at solution.
+
+    Notes
+    -----
+    If ``linesearch`` and ``trustregion`` are both specified, the line search
+    will be applied before the trust region.
+
+    """
+    # Check inputs
+    x = _check_x(x)
+    f = _check_f_min(f)
+    grad = _check_grad(grad, f, x)
+    H = _check_H(H, grad, x)
+    tol = _check_tol(tol)
+    maxiter = _check_maxiter(maxiter)
+    method = _check_method(method, grad, H, kwargs)
+    linesearch = _check_linesearch(linesearch, grad, H, lsopts)
+    trustregion = _check_trustregion(trustregion, grad, H, tropts)
+    # Minimization problem is equivalent to root-finding with
+    # cost = minimization target t, f == grad(t), and J == H(t).
+    result = _root_iteration(grad, x, H, tol, maxiter, method, linesearch, trustregion)
+    # Rename result keys for ``minimize`` output
+    result["grad"] = result.pop("f")
+    result["H"] = result.pop("J")
+    result["f"] = result.pop("cost")
+    # Return output dictionary
+    return result
+
+
+def _root_iteration(f, x, J, tol, maxiter, method, linesearch, trustregion):
+    r"""Run the iterations for ``root``."""
+    # Set up working data for direct methods
+    method.setup()
```

Comment:
Multi-secant and direct methods (i.e. not using any Jacobian) require the allocation of some internal, working data.  Method.setup() and Method.cleanup() at the end of this function ensure that this stuff is handled by anyone making a Method subclass.